File: macos/Sources/SuperVibe/AppState.swift
import Cocoa
import SwiftUI

final class AppState: @unchecked Sendable {

    // MARK: - State

    var stage: SessionStage = .idle
    var isRecording = false
    var statusText = "Ready"
    var asrText = ""
    var gptText = ""
    var polishEnabled: Bool = {
        if let val = AppState.loadConfigValue("polish_enabled") as? Bool {
            return val
        }
        return true  // default ON
    }() {
        didSet { saveConfig() }
    }
    var translationLanguage: String? = AppState.loadConfigValue("translation_language") as? String {
        didSet { saveConfig() }
    }
    private var sessionTranslate = false

    let vibeVoice = VibeVoiceSTT()
    let llmService = LLMService()
    private(set) var vibeVoiceAvailable = false

    var llmApiKey: String? {
        get { llmService.apiKey }
        set {
            llmService.apiKey = newValue
            saveConfig()
        }
    }

    var geminiApiKey: String? {
        get { llmService.geminiApiKey }
        set {
            llmService.geminiApiKey = newValue
            saveConfig()
        }
    }

    var llmProvider: LLMProvider {
        get { llmService.provider }
        set {
            llmService.provider = newValue
            saveConfig()
            onConfigChanged?()
        }
    }

    /// Called when isRecording changes (for menu bar icon update).
    var onRecordingChanged: ((Bool) -> Void)?
    /// Called when configuration changes (for menu rebuild).
    var onConfigChanged: (() -> Void)?
    /// Called when a user action fails and needs a visible alert.
    var onAlert: ((String, String) -> Void)?

    // MARK: - Private

    private let audioRecorder = AudioRecorder()
    private let hotkeyManager = HotkeyManager()
    private var overlayPanel: OverlayPanel?
    private let overlayVM = OverlayViewModel()
    private var processingTask: Task<Void, Never>?
    private var asrAccum = ""
    private var gptAccum = ""

    // MARK: - Init

    init() {
        loadConfig()
    }

    // MARK: - Hotkeys

    func setupHotkeys() {
        hotkeyManager.onToggleRecord = { [weak self] in
            DispatchQueue.main.async { self?.toggleRecording() }
        }
        hotkeyManager.onToggleTranslate = { [weak self] in
            DispatchQueue.main.async { self?.toggleTranslateRecording() }
        }
        hotkeyManager.onCancel = { [weak self] in
            DispatchQueue.main.async { self?.cancelSession() }
        }
        hotkeyManager.start()
    }

    // MARK: - Model Management

    func selectVibeModel(_ model: VibeVoiceSTT.Model) {
        vibeVoice.selectModel(model)
        saveConfig()
        log("[AppState] VibeVoice model: \(model.name)")

        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
            try? self?.vibeVoice.warmup()
        }

        onConfigChanged?()
    }

    func checkVibeVoiceAvailability() {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            let available = self?.vibeVoice.isAvailable() ?? false
            DispatchQueue.main.async {
                self?.vibeVoiceAvailable = available
                self?.onConfigChanged?()
            }
        }
    }

    // MARK: - Toggle

    func toggleRecording() {
        log("[AppState] toggleRecording -- isRecording=\(isRecording)")
        if isRecording { stopRecording() } else { startRecording(translate: false) }
    }

    func toggleTranslateRecording() {
        log("[AppState] toggleTranslateRecording -- isRecording=\(isRecording)")
        if isRecording {
            stopRecording()
        } else if translationLanguage != nil {
            if !llmService.isAvailable {
                log("[AppState] No LLM API key set, using raw transcription")
            }
            startRecording(translate: true)
        } else {
            log("[AppState] No translation language set")
            onAlert?("Translation Not Configured", "Please select a target language in the Translation menu before using the translate shortcut.")
        }
    }

    // MARK: - Start

    func startRecording(translate: Bool) {
        guard stage == .idle else {
            log("[AppState] startRecording skipped, stage=\(stage)")
            return
        }

        sessionTranslate = translate
        asrText = ""; gptText = ""
        asrAccum = ""; gptAccum = ""
        stage = .recording
        isRecording = true
        statusText = "Recording..."
        onRecordingChanged?(true)
        log("[AppState] Recording started -- mode=\(translate ? "translate(\(translationLanguage ?? "?"))" : "transcribe")")

        overlayVM.asrText = ""
        overlayVM.gptText = ""
        overlayVM.isTranslation = sessionTranslate
        overlayVM.stage = .recording
        showOverlay()

        // Audio level for waveform (both modes)
        audioRecorder.onAudioLevel = { [weak self] level in
            self?.overlayVM.pushLevel(level)
        }

        // Audio stays local; VibeVoice consumes the saved WAV after recording stops.
        audioRecorder.onAudioChunk = nil

        do {
            try audioRecorder.start()
        } catch {
            stage = .error; isRecording = false
            statusText = "Mic error: \(error.localizedDescription)"
            onRecordingChanged?(false)
            setOverlayStage(.error)
            log("[AppState] Mic error: \(error)")
            return
        }

        // Local processing happens on stop, after the WAV has been written.
    }

    // MARK: - Cancel (ESC)

    func cancelSession() {
        guard stage != .idle else { return }
        log("[AppState] Session cancelled (ESC)")
        stage = .idle
        statusText = "Ready"
        if isRecording {
            isRecording = false
            onRecordingChanged?(false)
            overlayVM.resetLevels()
            audioRecorder.stop()
        }
        // Cancel any running local processing.
        processingTask?.cancel()
        processingTask = nil
        hideOverlay()
    }

    // MARK: - Stop

    func stopRecording() {
        guard isRecording else { return }
        isRecording = false
        onRecordingChanged?(false)
        overlayVM.resetLevels()

        if stage == .recording {
            stage = .recognizing
            statusText = "Transcribing..."
            setOverlayStage(.recognizing)
        }

        log("[AppState] Recording stopped, starting local transcription")
        processingTask = Task { [weak self] in
            await self?.runLocalTranscription()
        }
    }

    // MARK: - Local Processing (VibeVoice)

    private func runLocalTranscription() async {
        do {
            let wavURL = try audioRecorder.stopAndSaveWAV()
            defer { try? FileManager.default.removeItem(at: wavURL) }

            log("[VibeVoice] Transcribing \(wavURL.path)...")
            let startTime = Date()
            let text = try vibeVoice.transcribe(wavPath: wavURL.path)
            let asrElapsed = Date().timeIntervalSince(startTime)
            log("[VibeVoice] ASR \(String(format: "%.1f", asrElapsed))s -- \(text.prefix(80))")

            guard !Task.isCancelled, stage != .idle else { return }

            if text.isEmpty {
                log("[VibeVoice] No speech detected")
                await MainActor.run {
                    scheduleReset(delay: 1.0)
                }
                return
            }

            await MainActor.run {
                asrText = text
                asrAccum = text
                setOverlayStage(.recognizing, asr: text)
            }

            // Determine if LLM post-processing is needed
            let needsTranslation = sessionTranslate && translationLanguage != nil
            let needsPolish = !sessionTranslate && polishEnabled
            let needsLLM = (needsTranslation || needsPolish) && llmService.isAvailable

            if needsLLM {
                await MainActor.run {
                    stage = .translating
                    if needsPolish {
                        statusText = "Polishing..."
                        overlayVM.isPolish = true
                    } else {
                        statusText = "Translating..."
                    }
                    setOverlayStage(.translating, asr: text)
                }

                guard !Task.isCancelled, stage != .idle else { return }

                let llmStart = Date()
                let result: String
                if needsTranslation, let lang = translationLanguage {
                    result = try await llmService.translate(text, to: lang, polish: polishEnabled)
                } else {
                    result = try await llmService.polish(text)
                }
                let llmElapsed = Date().timeIntervalSince(llmStart)
                log("[LLM] \(String(format: "%.1f", llmElapsed))s -- \(result.prefix(80))")

                guard !Task.isCancelled, stage != .idle else { return }

                await MainActor.run {
                    gptText = result
                    gptAccum = result
                    setOverlayStage(.translating, asr: text, gpt: result)
                }
            } else if needsTranslation && !llmService.isAvailable {
                log("[VibeVoice] Translation requested but no LLM API key set, using raw transcription")
            } else if needsPolish && !llmService.isAvailable {
                log("[VibeVoice] Polish enabled but no LLM API key set, using raw transcription")
            }

            guard !Task.isCancelled, stage != .idle else { return }

            await MainActor.run {
                finishSession()
            }
        } catch {
            guard !Task.isCancelled, stage != .idle else { return }
            await MainActor.run {
                log("[VibeVoice] Error: \(error)")
                stage = .error
                statusText = "Error: \(error.localizedDescription)"
                setOverlayStage(.error)
                scheduleReset(delay: 3.0)
            }
        }
    }

    // MARK: - Finish

    /// Strip trailing period for short single-sentence results.
    private func trimTrailingPunctuation(_ text: String) -> String {
        var s = text
        let midPunct: [Character] = [".", "\u{3002}", "!", "\u{FF01}", "?", "\u{FF1F}"]
        let body = s.dropLast()
        let hasMidPunct = body.contains(where: { midPunct.contains($0) })
        if !hasMidPunct, let last = s.last, last == "." || last == "\u{3002}" {
            s.removeLast()
        }
        return s
    }

    private func finishSession() {
        guard stage != .idle else { return }
        stage = .done

        // If GPT returned a refusal, discard it and use raw ASR text
        if !gptText.isEmpty && LLMService.looksLikeRefusal(gptText) {
            log("[AppState] GPT refusal detected, falling back to ASR text")
            gptText = ""
        }

        let raw = gptText.isEmpty ? asrText : gptText
        let finalText = trimTrailingPunctuation(raw)
        statusText = "Done"
        setOverlayStage(.done, asr: asrText, gpt: gptText)
        log("[AppState] Finish -- text=\"\(finalText.prefix(80))\"")

        guard !finalText.isEmpty else {
            scheduleReset(delay: 1.0)
            return
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.15) {
            PasteService.paste(finalText)
        }
        scheduleReset(delay: 1.5)
    }

    private func scheduleReset(delay: Double) {
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.hideOverlay()
            self?.stage = .idle
            self?.statusText = "Ready"
        }
    }

    // MARK: - Overlay

    private func setOverlayStage(
        _ stage: SessionStage,
        asr: String = "",
        gpt: String = ""
    ) {
        overlayVM.stage = stage
        overlayVM.asrText = asr
        overlayVM.gptText = gpt
        overlayVM.isTranslation = sessionTranslate
        // isPolish is set separately where needed (not reset here to avoid clearing it mid-session)
    }

    private func showOverlay() {
        if overlayPanel == nil {
            overlayVM.onCancel = { [weak self] in self?.cancelSession() }
            overlayVM.onConfirm = { [weak self] in self?.stopRecording() }
            overlayPanel = OverlayPanel()
            overlayPanel?.contentView = NSHostingView(
                rootView: OverlayRootView(vm: overlayVM)
            )
        }
        overlayPanel?.positionCenter()
        overlayPanel?.orderFront(nil)
    }

    private func hideOverlay() {
        overlayVM.resetLevels()
        overlayPanel?.contentView = nil   // tear down SwiftUI view tree (kills all animations & timers)
        overlayPanel?.orderOut(nil)
        overlayPanel = nil                // next showOverlay() will recreate
        overlayVM.stage = .idle
        overlayVM.asrText = ""
        overlayVM.gptText = ""
        overlayVM.isPolish = false
    }

    // MARK: - Persistence

    private static let configDir = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(".SuperVibe")
    private static let configFile = configDir.appendingPathComponent("config.json")

    static func loadConfigValue(_ key: String) -> Any? {
        guard let data = try? Data(contentsOf: configFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return json[key]
    }

    private func loadConfig() {
        guard let data = try? Data(contentsOf: Self.configFile),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }

        if let lang = json["translation_language"] as? String {
            translationLanguage = lang
        }
        if let modelId = json["vibevoice_model"] as? String,
           let model = VibeVoiceSTT.availableModels.first(where: { $0.id == modelId }) {
            vibeVoice.selectModel(model)
        }
        if let key = json["llm_api_key"] as? String, !key.isEmpty {
            llmService.apiKey = key
        }
        if let key = json["gemini_api_key"] as? String, !key.isEmpty {
            llmService.geminiApiKey = key
        }
        if let raw = json["llm_provider"] as? String, let p = LLMProvider(rawValue: raw) {
            llmService.provider = p
        }
        if let polish = json["polish_enabled"] as? Bool {
            polishEnabled = polish
        }
        if let modelId = json["llm_model"] as? String,
           let model = LLMService.claudeModels.first(where: { $0.id == modelId }) {
            llmService.selectedModel = model
        }
        if let modelId = json["gemini_model"] as? String,
           let model = LLMService.geminiModels.first(where: { $0.id == modelId }) {
            llmService.selectedGeminiModel = model
        }
    }

    func saveConfig() {
        let fm = FileManager.default
        try? fm.createDirectory(at: Self.configDir, withIntermediateDirectories: true)

        var config: [String: Any] = [:]
        if let data = try? Data(contentsOf: Self.configFile),
           let existing = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            config = existing
        }

        if let lang = translationLanguage {
            config["translation_language"] = lang
        } else {
            config.removeValue(forKey: "translation_language")
        }

        config.removeValue(forKey: "engine")
        config["vibevoice_model"] = vibeVoice.currentModel.id

        if let key = llmService.apiKey, !key.isEmpty {
            config["llm_api_key"] = key
        } else {
            config.removeValue(forKey: "llm_api_key")
        }

        if let key = llmService.geminiApiKey, !key.isEmpty {
            config["gemini_api_key"] = key
        } else {
            config.removeValue(forKey: "gemini_api_key")
        }

        config["llm_provider"] = llmService.provider.rawValue
        config["polish_enabled"] = polishEnabled
        config["llm_model"] = llmService.selectedModel.id
        config["gemini_model"] = llmService.selectedGeminiModel.id

        if let data = try? JSONSerialization.data(withJSONObject: config, options: [.prettyPrinted, .sortedKeys]) {
            try? data.write(to: Self.configFile)
        }
    }
}

File: SuperVibe/Services/AppState.cs
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Threading;
using SuperVibe.Models;

namespace SuperVibe.Services;

public class AppState : INotifyPropertyChanged, IDisposable
{
    private readonly ConfigService _configService = new();
    private AppConfig _config;

    // Services
    private readonly AudioRecorder _audioRecorder = new();
    private readonly WhisperSttService _whisperStt = new();
    private readonly LlmService _llmService = new();
    private readonly HotkeyManager _hotkeyManager = new();

    // State
    private SessionStage _stage = SessionStage.Idle;
    private string _statusText = "Ready";
    private string _asrText = "";
    private string _gptText = "";
    private bool _isRecording;
    private bool _sessionTranslate;
    private CancellationTokenSource? _processingCts;
    private Dispatcher? _dispatcher;

    public event PropertyChangedEventHandler? PropertyChanged;
    public event Action<bool>? RecordingChanged;
    public event Action? ConfigChanged;
    public event Action<string, string>? Alert;

    /// <summary>
    /// Called when the overlay should be updated. Args: stage, isTranslation, asrText, gptText, errorText
    /// </summary>
    public event Action<SessionStage, bool, string, string, string>? OverlayUpdate;

    /// <summary>
    /// Called when the overlay should be hidden.
    /// </summary>
    public event Action? OverlayHide;

    /// <summary>
    /// Called with audio RMS level for waveform.
    /// </summary>
    public event Action<float>? AudioLevelUpdate;

    public AppState()
    {
        _config = _configService.Load();
        SyncConfigToLlmService();
    }

    public void Initialize(Dispatcher dispatcher)
    {
        _dispatcher = dispatcher;

        // Wire up hotkeys — use BeginInvoke (not Invoke) to avoid blocking the
        // low-level keyboard hook callback, which has a strict timeout (~300ms).
        // Blocking via Invoke risks Windows silently removing the hook.
        _hotkeyManager.OnToggleRecord = () => dispatcher.BeginInvoke(ToggleRecording);
        _hotkeyManager.OnToggleTranslate = () => dispatcher.BeginInvoke(ToggleTranslateRecording);
        _hotkeyManager.OnCancel = () => dispatcher.BeginInvoke(Cancel);
        _hotkeyManager.Start();

        // Wire up audio level callback
        _audioRecorder.OnAudioLevel = level => AudioLevelUpdate?.Invoke(level);

        Debug.WriteLine("[AppState] Initialized");
    }

    // Properties

    public SessionStage Stage
    {
        get => _stage;
        private set { _stage = value; OnPropertyChanged(); OnPropertyChanged(nameof(IsIdle)); }
    }

    public string StatusText
    {
        get => _statusText;
        private set { _statusText = value; OnPropertyChanged(); }
    }

    public string AsrText
    {
        get => _asrText;
        private set { _asrText = value; OnPropertyChanged(); }
    }

    public string GptText
    {
        get => _gptText;
        private set { _gptText = value; OnPropertyChanged(); }
    }

    public bool IsRecording
    {
        get => _isRecording;
        private set { _isRecording = value; OnPropertyChanged(); RecordingChanged?.Invoke(value); }
    }

    public bool IsIdle => _stage == SessionStage.Idle;
    public AppConfig Config => _config;

    // Config accessors

    public bool PolishEnabled
    {
        get => _config.PolishEnabled;
        set { _config.PolishEnabled = value; SaveConfig(); OnPropertyChanged(); }
    }

    public string? TranslationLanguage
    {
        get => _config.TranslationLanguage;
        set { _config.TranslationLanguage = value; SaveConfig(); OnPropertyChanged(); }
    }

    public string LlmProvider
    {
        get => _config.LlmProvider;
        set
        {
            _config.LlmProvider = value;
            _llmService.Provider = value;
            SaveConfig(); OnPropertyChanged(); ConfigChanged?.Invoke();
        }
    }

    public string? LlmApiKey
    {
        get => _config.LlmApiKey;
        set
        {
            _config.LlmApiKey = string.IsNullOrEmpty(value) ? null : value;
            _llmService.ApiKey = _config.LlmApiKey;
            SaveConfig();
        }
    }

    public string? GeminiApiKey
    {
        get => _config.GeminiApiKey;
        set
        {
            _config.GeminiApiKey = string.IsNullOrEmpty(value) ? null : value;
            _llmService.GeminiApiKey = _config.GeminiApiKey;
            SaveConfig();
        }
    }

    public string SttModel
    {
        get => _config.SttModel;
        set { _config.SttModel = value; SaveConfig(); OnPropertyChanged(); ConfigChanged?.Invoke(); }
    }

    public string? LlmModel
    {
        get => _config.LlmModel;
        set
        {
            _config.LlmModel = value;
            if (value != null) _llmService.SelectModel(value);
            SaveConfig(); OnPropertyChanged(); ConfigChanged?.Invoke();
        }
    }

    public string? GeminiModel
    {
        get => _config.GeminiModel;
        set
        {
            _config.GeminiModel = value;
            if (value != null) _llmService.SelectModel(value);
            SaveConfig(); OnPropertyChanged(); ConfigChanged?.Invoke();
        }
    }

    // Session control — full pipeline

    public void ToggleRecording()
    {
        Debug.WriteLine($"[AppState] ToggleRecording -- isRecording={IsRecording}");
        if (IsRecording) StopSession(); else StartSession(false);
    }

    public void ToggleTranslateRecording()
    {
        Debug.WriteLine($"[AppState] ToggleTranslateRecording -- isRecording={IsRecording}");
        if (IsRecording)
        {
            StopSession();
        }
        else if (TranslationLanguage != null)
        {
            StartSession(true);
        }
        else
        {
            Debug.WriteLine("[AppState] No translation language set");
            Alert?.Invoke("Translation Not Configured", "Please select a target language in the Translation menu.");
        }
    }

    public void StartSession(bool translate)
    {
        if (Stage != SessionStage.Idle) return;

        _sessionTranslate = translate;
        AsrText = "";
        GptText = "";
        Stage = SessionStage.Recording;
        IsRecording = true;
        StatusText = "Recording...";
        Debug.WriteLine($"[AppState] Recording started -- mode={( translate ? $"translate({TranslationLanguage})" : "transcribe")}");

        OverlayUpdate?.Invoke(SessionStage.Recording, _sessionTranslate, "", "", "");

        try
        {
            _audioRecorder.Start();
        }
        catch (Exception ex)
        {
            Stage = SessionStage.Error;
            IsRecording = false;
            StatusText = $"Mic error: {ex.Message}";
            OverlayUpdate?.Invoke(SessionStage.Error, _sessionTranslate, "", "", ex.Message);
            ScheduleReset(3.0);
            Debug.WriteLine($"[AppState] Mic error: {ex}");
        }
    }

    public void StopSession()
    {
        if (!IsRecording) return;
        IsRecording = false;
        _audioRecorder.Stop();

        if (Stage == SessionStage.Recording)
        {
            Stage = SessionStage.Recognizing;
            StatusText = "Transcribing...";
            OverlayUpdate?.Invoke(SessionStage.Recognizing, _sessionTranslate, "", "", "");
        }

        Debug.WriteLine("[AppState] Recording stopped, starting transcription pipeline");
        _processingCts?.Cancel();
        _processingCts = new CancellationTokenSource();
        _ = RunTranscriptionPipeline(_processingCts.Token);
    }

    public void Cancel()
    {
        if (Stage == SessionStage.Idle) return;
        Debug.WriteLine("[AppState] Session cancelled (ESC)");

        _processingCts?.Cancel();
        _processingCts = null;

        if (IsRecording)
        {
            IsRecording = false;
            _audioRecorder.Stop();
        }

        Stage = SessionStage.Idle;
        StatusText = "Ready";
        OverlayHide?.Invoke();
    }

    // Pipeline

    private async Task RunTranscriptionPipeline(CancellationToken ct)
    {
        try
        {
            // Get audio buffer
            var audioBuffer = _audioRecorder.GetBufferAsFloat32();
            if (audioBuffer.Length == 0)
            {
                Debug.WriteLine("[AppState] No audio data");
                RunOnUI(() => ScheduleReset(1.0));
                return;
            }

            // Load Whisper model if needed
            await _whisperStt.LoadModelAsync(_config.SttModel, ct);
            if (ct.IsCancellationRequested) return;

            // Transcribe
            var sw = Stopwatch.StartNew();
            var text = await _whisperStt.TranscribeAsync(audioBuffer, ct);
            Debug.WriteLine($"[Whisper] ASR {sw.Elapsed.TotalSeconds:F1}s -- {text[..Math.Min(80, text.Length)]}");
            if (ct.IsCancellationRequested) return;

            if (string.IsNullOrWhiteSpace(text))
            {
                Debug.WriteLine("[Whisper] No speech detected");
                RunOnUI(() => ScheduleReset(1.0));
                return;
            }

            RunOnUI(() =>
            {
                AsrText = text;
                OverlayUpdate?.Invoke(SessionStage.Recognizing, _sessionTranslate, text, "", "");
            });

            // Determine if LLM post-processing is needed
            var needsTranslation = _sessionTranslate && TranslationLanguage != null;
            var needsPolish = !_sessionTranslate && PolishEnabled;
            var needsLlm = (needsTranslation || needsPolish) && _llmService.IsAvailable;

            if (needsLlm)
            {
                RunOnUI(() =>
                {
                    Stage = SessionStage.Translating;
                    StatusText = needsPolish ? "Polishing..." : "Translating...";
                    OverlayUpdate?.Invoke(SessionStage.Translating, _sessionTranslate, text, "", "");
                });

                if (ct.IsCancellationRequested) return;

                sw.Restart();
                string result;
                if (needsTranslation && TranslationLanguage != null)
                    result = await _llmService.TranslateAsync(text, TranslationLanguage, PolishEnabled);
                else
                    result = await _llmService.PolishAsync(text);

                Debug.WriteLine($"[LLM] {sw.Elapsed.TotalSeconds:F1}s -- {result[..Math.Min(80, result.Length)]}");
                if (ct.IsCancellationRequested) return;

                RunOnUI(() =>
                {
                    GptText = result;
                    OverlayUpdate?.Invoke(SessionStage.Translating, _sessionTranslate, text, result, "");
                });
            }

            if (ct.IsCancellationRequested) return;
            RunOnUI(FinishSession);
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            // User cancelled — Cancel() already handled cleanup
        }
        catch (OperationCanceledException ex)
        {
            // HTTP timeout or other non-user cancellation
            RunOnUI(() =>
            {
                Debug.WriteLine($"[AppState] Pipeline timeout: {ex.Message}");
                Stage = SessionStage.Error;
                StatusText = "Error: Request timed out";
                OverlayUpdate?.Invoke(SessionStage.Error, _sessionTranslate, "", "", "Request timed out");
                ScheduleReset(3.0);
            });
        }
        catch (Exception ex)
        {
            if (ct.IsCancellationRequested || Stage == SessionStage.Idle) return;
            RunOnUI(() =>
            {
                Debug.WriteLine($"[AppState] Pipeline error: {ex}");
                Stage = SessionStage.Error;
                StatusText = $"Error: {ex.Message}";
                OverlayUpdate?.Invoke(SessionStage.Error, _sessionTranslate, "", "", ex.Message);
                ScheduleReset(3.0);
            });
        }
    }

    private void FinishSession()
    {
        if (Stage == SessionStage.Idle) return;
        Stage = SessionStage.Done;

        // Refusal detection
        if (!string.IsNullOrEmpty(GptText) && LlmService.LooksLikeRefusal(GptText))
        {
            Debug.WriteLine("[AppState] GPT refusal detected, falling back to ASR text");
            GptText = "";
        }

        var raw = string.IsNullOrEmpty(GptText) ? AsrText : GptText;
        var finalText = TrimTrailingPunctuation(raw);
        StatusText = "Done";
        OverlayUpdate?.Invoke(SessionStage.Done, _sessionTranslate, AsrText, GptText, "");
        Debug.WriteLine($"[AppState] Finish -- text=\"{finalText[..Math.Min(80, finalText.Length)]}\"");

        if (string.IsNullOrEmpty(finalText))
        {
            ScheduleReset(1.0);
            return;
        }

        // Paste after short delay
        var timer = new DispatcherTimer { Interval = TimeSpan.FromMilliseconds(150) };
        timer.Tick += (_, _) =>
        {
            timer.Stop();
            try
            {
                ClipboardPasteService.Paste(finalText);
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[AppState] Paste failed: {ex.Message}");
            }
        };
        timer.Start();

        ScheduleReset(1.5);
    }

    private static string TrimTrailingPunctuation(string text)
    {
        if (string.IsNullOrEmpty(text)) return text;
        var s = text;
        char[] midPunct = ['.', '\u3002', '!', '\uFF01', '?', '\uFF1F'];
        var body = s[..^1];
        var hasMidPunct = body.Any(c => midPunct.Contains(c));
        if (!hasMidPunct && s.Length > 0)
        {
            var last = s[^1];
            if (last == '.' || last == '\u3002')
                s = s[..^1];
        }
        return s;
    }

    private void ScheduleReset(double delaySec)
    {
        var timer = new DispatcherTimer { Interval = TimeSpan.FromSeconds(delaySec) };
        timer.Tick += (_, _) =>
        {
            timer.Stop();
            OverlayHide?.Invoke();
            Stage = SessionStage.Idle;
            StatusText = "Ready";
        };
        timer.Start();
    }

    private void RunOnUI(Action action)
    {
        if (_dispatcher != null)
            _dispatcher.Invoke(action);
        else
            action();
    }

    private void SyncConfigToLlmService()
    {
        _llmService.Provider = _config.LlmProvider;
        _llmService.ApiKey = _config.LlmApiKey;
        _llmService.GeminiApiKey = _config.GeminiApiKey;

        if (_config.LlmModel != null)
        {
            var model = LlmModels.ClaudeModels.FirstOrDefault(m => m.Id == _config.LlmModel);
            if (model != null) _llmService.SelectedClaudeModel = model;
        }
        if (_config.GeminiModel != null)
        {
            var model = LlmModels.GeminiModels.FirstOrDefault(m => m.Id == _config.GeminiModel);
            if (model != null) _llmService.SelectedGeminiModel = model;
        }
    }

    private void SaveConfig()
    {
        _configService.Save(_config);
    }

    protected void OnPropertyChanged([CallerMemberName] string? name = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
    }

    public void Dispose()
    {
        _processingCts?.Cancel();
        _processingCts = null;
        _hotkeyManager.Dispose();
        _audioRecorder.Dispose();
        _whisperStt.Dispose();
    }
}
